### PR TITLE
improve error message

### PR DESCRIPTION
**Pull request by oeway**
When developing python plugin with a remote plugin engine, we cannot see much error information. This PR will improve the situation by sending traceback stack to ImJoy.

Tested with python 2 and 3

### CODE REVIEW

**Comment by MartinHjelmare** [comment]
Note that this PR will have merge conflict with https://github.com/oeway/ImJoy-Engine/pull/30.

File: imjoy/imjoyUtils.py [hunk 1]
```diff
@@ -153,12 +152,11 @@ def task_worker(self, q, logger, abort):
                             result = method(*args)
                             resolve(result)
                         except Exception as e:
+                            traceback_error = traceback.format_exc()
                             logger.error(
```
Comment (MartinHjelmare) [inline imjoy/imjoyUtils.py]: If we want to log the traceback, we can use `logger.exception`.

File: imjoy/imjoyUtils.py [hunk 2]
```diff
@@ -135,10 +135,9 @@ def task_worker(self, q, logger, abort):
                             raise Exception("unsupported type")
                         self.emit({"type": "executeSuccess"})
                     except Exception as e:
-                        logger.info(
-                            "error during execution: %s", traceback.format_exc()
-                        )
-                        self.emit({"type": "executeFailure", "error": repr(e)})
+                        traceback_error = traceback.format_exc()
+                        logger.info("error during execution: %s", traceback_error)
```
Comment (MartinHjelmare) [inline imjoy/imjoyUtils.py]: Should it be info level?

Comment (oeway) [reply imjoy/imjoyUtils.py]: No, I will change.

**Comment by oeway** [comment]
Should me merge #31  first, I can comment on #31 and we can then cancel this one.

**Comment by MartinHjelmare** [comment]
Do you mean merge #30 first?

**Comment by oeway** [comment]
Yes, a quick question, this PR aims to pass all necessary error information to the ImJoy web interface, we can either directly pass `traceback.format_exc()` to the frontend, or if there is better way to send more error information, we can also send an dictionary with for example, error name, line number, traceback etc to ImJoy, we can then pass the error information to show in the ImJoy interface. Potentially, we can highlight the error line in the plugin code editor, and this will help a lot for debugging.

File: imjoy/imjoyUtils.py (hunk 1 above)
Comment (oeway) [reply imjoy/imjoyUtils.py]: Sounds good. And one thing we need to think is also how do we actually use these logs.

Ideally, in the future, we should have a way to collect these logs and expose to the ImJoy app, and each plugin, and also the engine should have a place to read these logs, and depending on different types, we can apply css styles (e.g.: error in red) to show a log window.

**Comment by oeway** [comment]
Test code:
```
from imjoy import api

def lumberjack():
    bright_side_of_death()

def bright_side_of_death():
    return tuple()[0]

class ImJoyPlugin():
    def setup(self):
        api.log('initialized')

    def run(self, ctx):
        lumberjack()
 
api.export(ImJoyPlugin())
```

The traceback sent to the front end was:
```
Error: Traceback (most recent call last): File "/Users/weiouyang/workspace/ImJoy-Engine/imjoy/imjoyUtils3.py", line 60, in task_worker result = method(*args) File "<string>", line 15, in run File "<string>", line 5, in lumberjack File "<string>", line 8, in bright_side_of_death IndexError: tuple index out of range
```
This is generated by `traceback.format_exc()`, we just need to remove second line and replace `File "<string>"` to `Plugin "{PLUGIN_NAME}"`, such that we will get:
```
Error: Traceback (most recent call last):
Plugin "{PLUGIN_NAME}", line 15, in run 
Plugin "{PLUGIN_NAME}", line 5, in lumberjack
File "<string>", line 8, in bright_side_of_death
IndexError: tuple index out of range
```

A further improvement would be calculate the line number offset such that we will have the actual line number.

File: imjoy/imjoyPluginEngine.py [hunk 3]
```diff
@@ -825,6 +825,9 @@ def connect(sid, environ):
                 logger.debug("message from %s", pid)
                 if kwargs["type"] == "initialized":
                     addPlugin(plugin_info, sid)
+                elif kwargs["type"] == "executeFailure":
+                    logger.info("Killing plugin %s due to exeuction failure.", pid)
+                    killPlugin(pid)
```
Comment (oeway) [inline imjoy/imjoyPluginEngine.py]: I added that to make sure that we kill the plugin when an executeFailure happened.

File: imjoy/imjoyUtils3.py [hunk 4]
```diff
@@ -41,10 +41,10 @@
                             raise Exception("unsupported type")
                         self.emit({"type": "executeSuccess"})
                     except Exception as e:
-                        logger.info(
-                            "error during execution: %s", traceback.format_exc()
-                        )
-                        self.emit({"type": "executeFailure", "error": repr(e)})
+                        traceback_error = traceback.format_exc()
+                        logger.error("error during execution: %s", traceback_error)
+                        self.emit({"type": "executeFailure", "error": traceback_error})
```
Comment (oeway) [inline imjoy/imjoyUtils3.py]: Notice that here we can only emit a string (not Exception)

File: imjoy/imjoyUtils3.py [hunk 5]
```diff
@@ -58,12 +58,11 @@
...
                         except Exception as e:
+                            traceback_error = traceback.format_exc()
                             logger.error(
-                                "error in method %s: %s",
-                                d["name"],
-                                traceback.format_exc(),
+                                "error in method %s: %s", d["name"], traceback_error
                             )
-                            reject(e)
+                            reject(Exception(formatTraceback(traceback_error)))
```
Comment (oeway) [inline imjoy/imjoyUtils3.py]: The error string will be formatted and convert into an Exception, the reject function will encode it and send it as an Error object.

**Pull request merged**